{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f92e52b1-dab6-4384-a01f-0b3d55f75002",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7be2cae1-0442-4a2a-8b89-a8526c55878b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eca7ca81-a134-4082-9aab-a754bb2d2765",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "70f83569-d73c-4486-9afa-0edfdb707f4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATESTR = '220420'\n",
    "NBNAME = DATESTR + '-pt-table-results-comparison'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "35931c87-9560-43bf-b940-54a0985e2074",
   "metadata": {},
   "outputs": [],
   "source": [
    "infiles = dict(\n",
    "    david_table=Path('data-src/220420-midas-pt-table.csv'),\n",
    "    my_table=Path('data-processed/220419-set3-files-list/220419-200726-gold-standard-files.csv'),\n",
    "    gambit=Path('data-src/220420-query-1.0-beta2.csv'),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e3998319-4fa6-4d94-965b-d858f5cc47d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "processed_out = Path('data-processed') / NBNAME\n",
    "processed_out.mkdir(exist_ok=True)\n",
    "\n",
    "outfiles = dict(\n",
    "    comp=processed_out / f'{DATESTR}-pt-table-results-comp.csv',\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66488197-9b55-481c-9c34-c109bfd1c34c",
   "metadata": {},
   "source": [
    "## Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "333beafe-a557-4b33-8a11-9da3def69464",
   "metadata": {},
   "outputs": [],
   "source": [
    "def yesno_to_bool(x):\n",
    "    if x == 'yes':\n",
    "        return True\n",
    "    elif x == 'no':\n",
    "        return False\n",
    "    raise ValueError(x)\n",
    "    \n",
    "\n",
    "def stripprefix(prefix, s, strict=True):\n",
    "    if s.startswith(prefix):\n",
    "        return s[len(prefix):]\n",
    "    if strict:\n",
    "        raise ValueError('String does not contain prefix')\n",
    "    return s\n",
    "    \n",
    "\n",
    "def stripsuffix(suffix, s, strict=True):\n",
    "    if s.endswith(suffix):\n",
    "        return s[:-len(suffix)]\n",
    "    if strict:\n",
    "        raise ValueError('String does not contain suffix')\n",
    "    return s"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baa3a378-6f94-46d1-a2ed-94e95b20e11b",
   "metadata": {},
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6372468b-702d-456d-b344-c2f267bf78d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_df = pd.read_csv(infiles['my_table'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "259579ef-4126-49eb-9b1b-31636f404b19",
   "metadata": {},
   "source": [
    "### David"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "069e0ec1-e5a1-4dca-a854-a25e9ef1defa",
   "metadata": {},
   "outputs": [],
   "source": [
    "david = pd.read_csv(infiles['david_table'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9d24e307-5b2f-4d89-a6aa-4cd75116d736",
   "metadata": {},
   "outputs": [],
   "source": [
    "del david['jared_id needs edit']\n",
    "david = david.set_index('jared_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "2258eb51-b0dc-4569-bd29-21dafcc5d3c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "david.columns = ['wgs_date', 'class', 'short_name', 'fastq_name', 'gambit_dist', 'gambit_sp_thresh', 'has_16s', 'gambit_genus', 'gambit_species', 'reported', 'cap']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8adc794e-bdca-4c43-a446-eace0f2c81aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "david['has_16s'] = david['has_16s'].map(yesno_to_bool)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "684e9abe-5b7e-4d59-9132-3131cdcc3ee8",
   "metadata": {},
   "outputs": [],
   "source": [
    "david.loc[david['gambit_species'] == 'none', 'gambit_species'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "77a05511-5b4a-4b1b-996f-5a616438f659",
   "metadata": {},
   "outputs": [],
   "source": [
    "david.loc[david['gambit_genus'] == 'none', 'gambit_genus'] = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "182d2423-5902-44c9-add8-49bbcda43c1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "david['gambit_predicted'] = [row.gambit_genus if row.gambit_species is None else ' '.join([row.gambit_genus, row.gambit_species]) for _, row in david.iterrows()]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6cd470b-1556-472e-b3d2-efa8bf707ad9",
   "metadata": {},
   "source": [
    "### GAMBIT results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "06559cf4-2fc1-426c-91bd-e066251eba1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "gambit = pd.read_csv(infiles['gambit'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "87acb035-db70-48cf-9075-b4a830b562b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gambit = gambit.set_index('query.name')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bffe2d46-38f1-4cb2-85d8-4d3c1a5ef8a2",
   "metadata": {},
   "source": [
    "## Inspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4825af40-9224-4f11-bae9-4b4e974d5389",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "class\n",
       "BACT1_2016    6\n",
       "BACT1_2017    6\n",
       "BACT1_2018    6\n",
       "BACT1_2019    6\n",
       "BACT1_2020    7\n",
       "BACT2_2017    7\n",
       "BACT2_2018    7\n",
       "BACT2_2019    8\n",
       "BACT3_2017    7\n",
       "BACT3_2018    8\n",
       "BACT3_2019    5\n",
       "LPX1 2017     3\n",
       "LPX1_2018     3\n",
       "LPX1_2019     3\n",
       "LPX2_2018     3\n",
       "LPX2_2019     3\n",
       "dtype: int64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "david.groupby('class').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c0226068-211b-405f-a7a5-6e5537be0fe7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_date</th>\n",
       "      <th>class</th>\n",
       "      <th>short_name</th>\n",
       "      <th>fastq_name</th>\n",
       "      <th>gambit_dist</th>\n",
       "      <th>gambit_sp_thresh</th>\n",
       "      <th>has_16s</th>\n",
       "      <th>gambit_genus</th>\n",
       "      <th>gambit_species</th>\n",
       "      <th>reported</th>\n",
       "      <th>cap</th>\n",
       "      <th>gambit_predicted</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>jared_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17AC0006313-1</th>\n",
       "      <td>6/29/17</td>\n",
       "      <td>BACT2_2017</td>\n",
       "      <td>17AC0006313-1_S23.fasta</td>\n",
       "      <td>17AC0006313-1_S23_L001_R1_001.fastq</td>\n",
       "      <td>0.913</td>\n",
       "      <td>0.944</td>\n",
       "      <td>True</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>None</td>\n",
       "      <td>Veilonella genus with GAMBIT, V. parvula with 16S</td>\n",
       "      <td>Mixed Sample 2: Staphylococcus hominis and Vei...</td>\n",
       "      <td>Veillonella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18AC0018937-BAP</th>\n",
       "      <td>10/25/18</td>\n",
       "      <td>BACT3_2018</td>\n",
       "      <td>18AC0018937-BAP</td>\n",
       "      <td>18AC0018937-BAP_S11_L001_R1_001.fastq</td>\n",
       "      <td>0.942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>Proteus sp / Proteus vulgaris</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19AC0011210</th>\n",
       "      <td>7/9/19</td>\n",
       "      <td>BACT2_2019</td>\n",
       "      <td>19AC0011210-W1</td>\n",
       "      <td>19AC0011210_S5_L001_R1_001.fastq</td>\n",
       "      <td>0.992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Granulicatella adiacens</td>\n",
       "      <td>Granulicatella sp / Granulicatella adiacens</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19AC0016708</th>\n",
       "      <td>10/17/19</td>\n",
       "      <td>BACT3_2019</td>\n",
       "      <td>19AC0016708</td>\n",
       "      <td>19AC0016708_S8_L001_R1_001.fastq</td>\n",
       "      <td>0.325</td>\n",
       "      <td>0.921</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>None</td>\n",
       "      <td>Pseudomonas putida group (species reported wit...</td>\n",
       "      <td>Pseudomonas sp/Pseudomonas putida /Pseudomonas...</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 wgs_date       class               short_name  \\\n",
       "jared_id                                                         \n",
       "17AC0006313-1     6/29/17  BACT2_2017  17AC0006313-1_S23.fasta   \n",
       "18AC0018937-BAP  10/25/18  BACT3_2018          18AC0018937-BAP   \n",
       "19AC0011210        7/9/19  BACT2_2019           19AC0011210-W1   \n",
       "19AC0016708      10/17/19  BACT3_2019              19AC0016708   \n",
       "\n",
       "                                            fastq_name  gambit_dist  \\\n",
       "jared_id                                                              \n",
       "17AC0006313-1      17AC0006313-1_S23_L001_R1_001.fastq        0.913   \n",
       "18AC0018937-BAP  18AC0018937-BAP_S11_L001_R1_001.fastq        0.942   \n",
       "19AC0011210           19AC0011210_S5_L001_R1_001.fastq        0.992   \n",
       "19AC0016708           19AC0016708_S8_L001_R1_001.fastq        0.325   \n",
       "\n",
       "                 gambit_sp_thresh  has_16s gambit_genus gambit_species  \\\n",
       "jared_id                                                                 \n",
       "17AC0006313-1               0.944     True  Veillonella           None   \n",
       "18AC0018937-BAP               NaN     True         None           None   \n",
       "19AC0011210                   NaN     True         None           None   \n",
       "19AC0016708                 0.921     True  Pseudomonas           None   \n",
       "\n",
       "                                                          reported  \\\n",
       "jared_id                                                             \n",
       "17AC0006313-1    Veilonella genus with GAMBIT, V. parvula with 16S   \n",
       "18AC0018937-BAP                                   Proteus vulgaris   \n",
       "19AC0011210                                Granulicatella adiacens   \n",
       "19AC0016708      Pseudomonas putida group (species reported wit...   \n",
       "\n",
       "                                                               cap  \\\n",
       "jared_id                                                             \n",
       "17AC0006313-1    Mixed Sample 2: Staphylococcus hominis and Vei...   \n",
       "18AC0018937-BAP                      Proteus sp / Proteus vulgaris   \n",
       "19AC0011210            Granulicatella sp / Granulicatella adiacens   \n",
       "19AC0016708      Pseudomonas sp/Pseudomonas putida /Pseudomonas...   \n",
       "\n",
       "                gambit_predicted  \n",
       "jared_id                          \n",
       "17AC0006313-1        Veillonella  \n",
       "18AC0018937-BAP             None  \n",
       "19AC0011210                 None  \n",
       "19AC0016708          Pseudomonas  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "david[david['has_16s']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c7e053fa-5a4e-4ea5-be12-80bf756c52e8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wgs_date</th>\n",
       "      <th>class</th>\n",
       "      <th>short_name</th>\n",
       "      <th>fastq_name</th>\n",
       "      <th>gambit_dist</th>\n",
       "      <th>gambit_sp_thresh</th>\n",
       "      <th>has_16s</th>\n",
       "      <th>gambit_genus</th>\n",
       "      <th>gambit_species</th>\n",
       "      <th>reported</th>\n",
       "      <th>cap</th>\n",
       "      <th>gambit_predicted</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>jared_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17AC0001410</th>\n",
       "      <td>2/25/17</td>\n",
       "      <td>BACT1_2017</td>\n",
       "      <td>17AC0001410.fasta</td>\n",
       "      <td>17AC0001410_S9_L001_R1_001.fastq</td>\n",
       "      <td>0.161</td>\n",
       "      <td>0.347</td>\n",
       "      <td>False</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>coli</td>\n",
       "      <td>Negative for E coli O157/Aeromonas/Campylobact...</td>\n",
       "      <td>Mixed Sample 1: Exclude E coli O157/Aeromonas/...</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17AC0001410A</th>\n",
       "      <td>3/4/17</td>\n",
       "      <td>BACT1_2017</td>\n",
       "      <td>17AC0001410A.fasta</td>\n",
       "      <td>17AC0001410A_S12_L001_R1_001.fastq</td>\n",
       "      <td>0.131</td>\n",
       "      <td>0.470</td>\n",
       "      <td>False</td>\n",
       "      <td>Enterococcus</td>\n",
       "      <td>faecalis</td>\n",
       "      <td>Negative for E coli O157/Aeromonas/Campylobact...</td>\n",
       "      <td>Mixed Sample 1: Exclude E coli O157/Aeromonas/...</td>\n",
       "      <td>Enterococcus faecalis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17AC0006313-1</th>\n",
       "      <td>6/29/17</td>\n",
       "      <td>BACT2_2017</td>\n",
       "      <td>17AC0006313-1_S23.fasta</td>\n",
       "      <td>17AC0006313-1_S23_L001_R1_001.fastq</td>\n",
       "      <td>0.913</td>\n",
       "      <td>0.944</td>\n",
       "      <td>True</td>\n",
       "      <td>Veillonella</td>\n",
       "      <td>None</td>\n",
       "      <td>Veilonella genus with GAMBIT, V. parvula with 16S</td>\n",
       "      <td>Mixed Sample 2: Staphylococcus hominis and Vei...</td>\n",
       "      <td>Veillonella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17AC0012453</th>\n",
       "      <td>10/26/17</td>\n",
       "      <td>BACT3_2017</td>\n",
       "      <td>17AC0012453_S1.fasta</td>\n",
       "      <td>17AC0012453_S1_L001_R1_001.fastq</td>\n",
       "      <td>0.886</td>\n",
       "      <td>0.928</td>\n",
       "      <td>False</td>\n",
       "      <td>Corynebacterium</td>\n",
       "      <td>None</td>\n",
       "      <td>Corynebacterium sp</td>\n",
       "      <td>Corynebacterium sp /Corynebacterium striatum</td>\n",
       "      <td>Corynebacterium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17AC0012454</th>\n",
       "      <td>10/26/17</td>\n",
       "      <td>BACT3_2017</td>\n",
       "      <td>17AC0012454_S2.fasta</td>\n",
       "      <td>17AC0012454_S2_L001_R1_001.fastq</td>\n",
       "      <td>0.019</td>\n",
       "      <td>0.937</td>\n",
       "      <td>False</td>\n",
       "      <td>Aerococcus</td>\n",
       "      <td>None</td>\n",
       "      <td>Aerococcus sp</td>\n",
       "      <td>Aerococcus sp/ Aerococcus viridans</td>\n",
       "      <td>Aerococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18AC0002826</th>\n",
       "      <td>2/23/18</td>\n",
       "      <td>BACT1_2018</td>\n",
       "      <td>18AC0002826</td>\n",
       "      <td>18AC0002826_S11_L001_R1_001.fastq</td>\n",
       "      <td>0.016</td>\n",
       "      <td>0.469</td>\n",
       "      <td>False</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>agalactiae</td>\n",
       "      <td>Streptococcus agalactiae group B /Finegoldia m...</td>\n",
       "      <td>Streptococcus agalactiae group B /Finegoldia m...</td>\n",
       "      <td>Streptococcus agalactiae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18AC0018937-BAP</th>\n",
       "      <td>10/25/18</td>\n",
       "      <td>BACT3_2018</td>\n",
       "      <td>18AC0018937-BAP</td>\n",
       "      <td>18AC0018937-BAP_S11_L001_R1_001.fastq</td>\n",
       "      <td>0.942</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Proteus vulgaris</td>\n",
       "      <td>Proteus sp / Proteus vulgaris</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19AC0011210</th>\n",
       "      <td>7/9/19</td>\n",
       "      <td>BACT2_2019</td>\n",
       "      <td>19AC0011210-W1</td>\n",
       "      <td>19AC0011210_S5_L001_R1_001.fastq</td>\n",
       "      <td>0.992</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Granulicatella adiacens</td>\n",
       "      <td>Granulicatella sp / Granulicatella adiacens</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19AC0016708</th>\n",
       "      <td>10/17/19</td>\n",
       "      <td>BACT3_2019</td>\n",
       "      <td>19AC0016708</td>\n",
       "      <td>19AC0016708_S8_L001_R1_001.fastq</td>\n",
       "      <td>0.325</td>\n",
       "      <td>0.921</td>\n",
       "      <td>True</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>None</td>\n",
       "      <td>Pseudomonas putida group (species reported wit...</td>\n",
       "      <td>Pseudomonas sp/Pseudomonas putida /Pseudomonas...</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19AC0016710</th>\n",
       "      <td>10/17/19</td>\n",
       "      <td>BACT3_2019</td>\n",
       "      <td>19AC0016710</td>\n",
       "      <td>19AC0016710_S10_L001_R1_001.fastq</td>\n",
       "      <td>0.039</td>\n",
       "      <td>0.522</td>\n",
       "      <td>False</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>pyogenes</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 wgs_date       class               short_name  \\\n",
       "jared_id                                                         \n",
       "17AC0001410       2/25/17  BACT1_2017        17AC0001410.fasta   \n",
       "17AC0001410A       3/4/17  BACT1_2017       17AC0001410A.fasta   \n",
       "17AC0006313-1     6/29/17  BACT2_2017  17AC0006313-1_S23.fasta   \n",
       "17AC0012453      10/26/17  BACT3_2017     17AC0012453_S1.fasta   \n",
       "17AC0012454      10/26/17  BACT3_2017     17AC0012454_S2.fasta   \n",
       "18AC0002826       2/23/18  BACT1_2018              18AC0002826   \n",
       "18AC0018937-BAP  10/25/18  BACT3_2018          18AC0018937-BAP   \n",
       "19AC0011210        7/9/19  BACT2_2019           19AC0011210-W1   \n",
       "19AC0016708      10/17/19  BACT3_2019              19AC0016708   \n",
       "19AC0016710      10/17/19  BACT3_2019              19AC0016710   \n",
       "\n",
       "                                            fastq_name  gambit_dist  \\\n",
       "jared_id                                                              \n",
       "17AC0001410           17AC0001410_S9_L001_R1_001.fastq        0.161   \n",
       "17AC0001410A        17AC0001410A_S12_L001_R1_001.fastq        0.131   \n",
       "17AC0006313-1      17AC0006313-1_S23_L001_R1_001.fastq        0.913   \n",
       "17AC0012453           17AC0012453_S1_L001_R1_001.fastq        0.886   \n",
       "17AC0012454           17AC0012454_S2_L001_R1_001.fastq        0.019   \n",
       "18AC0002826          18AC0002826_S11_L001_R1_001.fastq        0.016   \n",
       "18AC0018937-BAP  18AC0018937-BAP_S11_L001_R1_001.fastq        0.942   \n",
       "19AC0011210           19AC0011210_S5_L001_R1_001.fastq        0.992   \n",
       "19AC0016708           19AC0016708_S8_L001_R1_001.fastq        0.325   \n",
       "19AC0016710          19AC0016710_S10_L001_R1_001.fastq        0.039   \n",
       "\n",
       "                 gambit_sp_thresh  has_16s     gambit_genus gambit_species  \\\n",
       "jared_id                                                                     \n",
       "17AC0001410                 0.347    False      Escherichia           coli   \n",
       "17AC0001410A                0.470    False     Enterococcus       faecalis   \n",
       "17AC0006313-1               0.944     True      Veillonella           None   \n",
       "17AC0012453                 0.928    False  Corynebacterium           None   \n",
       "17AC0012454                 0.937    False       Aerococcus           None   \n",
       "18AC0002826                 0.469    False    Streptococcus     agalactiae   \n",
       "18AC0018937-BAP               NaN     True             None           None   \n",
       "19AC0011210                   NaN     True             None           None   \n",
       "19AC0016708                 0.921     True      Pseudomonas           None   \n",
       "19AC0016710                 0.522    False    Streptococcus       pyogenes   \n",
       "\n",
       "                                                          reported  \\\n",
       "jared_id                                                             \n",
       "17AC0001410      Negative for E coli O157/Aeromonas/Campylobact...   \n",
       "17AC0001410A     Negative for E coli O157/Aeromonas/Campylobact...   \n",
       "17AC0006313-1    Veilonella genus with GAMBIT, V. parvula with 16S   \n",
       "17AC0012453                                     Corynebacterium sp   \n",
       "17AC0012454                                          Aerococcus sp   \n",
       "18AC0002826      Streptococcus agalactiae group B /Finegoldia m...   \n",
       "18AC0018937-BAP                                   Proteus vulgaris   \n",
       "19AC0011210                                Granulicatella adiacens   \n",
       "19AC0016708      Pseudomonas putida group (species reported wit...   \n",
       "19AC0016710                                Streptococcus pyogenes    \n",
       "\n",
       "                                                               cap  \\\n",
       "jared_id                                                             \n",
       "17AC0001410      Mixed Sample 1: Exclude E coli O157/Aeromonas/...   \n",
       "17AC0001410A     Mixed Sample 1: Exclude E coli O157/Aeromonas/...   \n",
       "17AC0006313-1    Mixed Sample 2: Staphylococcus hominis and Vei...   \n",
       "17AC0012453           Corynebacterium sp /Corynebacterium striatum   \n",
       "17AC0012454                     Aerococcus sp/ Aerococcus viridans   \n",
       "18AC0002826      Streptococcus agalactiae group B /Finegoldia m...   \n",
       "18AC0018937-BAP                      Proteus sp / Proteus vulgaris   \n",
       "19AC0011210            Granulicatella sp / Granulicatella adiacens   \n",
       "19AC0016708      Pseudomonas sp/Pseudomonas putida /Pseudomonas...   \n",
       "19AC0016710                                Streptococcus pyogenes    \n",
       "\n",
       "                         gambit_predicted  \n",
       "jared_id                                   \n",
       "17AC0001410              Escherichia coli  \n",
       "17AC0001410A        Enterococcus faecalis  \n",
       "17AC0006313-1                 Veillonella  \n",
       "17AC0012453               Corynebacterium  \n",
       "17AC0012454                    Aerococcus  \n",
       "18AC0002826      Streptococcus agalactiae  \n",
       "18AC0018937-BAP                      None  \n",
       "19AC0011210                          None  \n",
       "19AC0016708                   Pseudomonas  \n",
       "19AC0016710        Streptococcus pyogenes  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "david[(david['gambit_genus'] + ' ' + david['gambit_species']) != david['reported']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da02f11b-a6cd-49a9-996d-faab044543e2",
   "metadata": {},
   "source": [
    "## Compare to latest GAMBIT results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7bf68e73-ba13-4603-aefd-83a571c1e1ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "left = david[['gambit_dist', 'gambit_sp_thresh', 'gambit_genus', 'gambit_species', 'reported', 'cap', 'gambit_predicted']]\n",
    "left.columns = ['pt_' + n for n in left.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "1dc52f82-0c3b-4619-b2ec-f75784efda4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "right = gambit.copy()\n",
    "del right['query.path']\n",
    "right.columns = ['new_' + n for n in right.columns]\n",
    "right.index.name = 'jared_id'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "4ab9baca-ed57-41c5-ab2a-e73e40f4913a",
   "metadata": {},
   "outputs": [],
   "source": [
    "comp = left.join(right, on='jared_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9f5ee13d-f268-49ff-a3a8-95bce8834783",
   "metadata": {},
   "outputs": [],
   "source": [
    "matches = comp['new_predicted.name'] == comp['pt_gambit_predicted']\n",
    "matches |= pd.isnull(comp['new_predicted.name']) & pd.isnull(comp['pt_gambit_predicted'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "9ef7212a-6ae0-46b7-be71-932dcee2e0b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pt_gambit_dist</th>\n",
       "      <th>pt_gambit_sp_thresh</th>\n",
       "      <th>pt_gambit_genus</th>\n",
       "      <th>pt_gambit_species</th>\n",
       "      <th>pt_reported</th>\n",
       "      <th>pt_cap</th>\n",
       "      <th>pt_gambit_predicted</th>\n",
       "      <th>new_predicted.name</th>\n",
       "      <th>new_predicted.rank</th>\n",
       "      <th>new_predicted.ncbi_id</th>\n",
       "      <th>new_predicted.threshold</th>\n",
       "      <th>new_closest.distance</th>\n",
       "      <th>new_closest.description</th>\n",
       "      <th>new_next.name</th>\n",
       "      <th>new_next.rank</th>\n",
       "      <th>new_next.ncbi_id</th>\n",
       "      <th>new_next.threshold</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>jared_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16AC1611140BCAP</th>\n",
       "      <td>0.337</td>\n",
       "      <td>0.554</td>\n",
       "      <td>Propionibacterium (Cutibacterium)</td>\n",
       "      <td>acnes</td>\n",
       "      <td>Propionibacterium (Cutibacterium) acnes</td>\n",
       "      <td>Propionibacterium</td>\n",
       "      <td>Propionibacterium (Cutibacterium) acnes</td>\n",
       "      <td>Cutibacterium acnes</td>\n",
       "      <td>species</td>\n",
       "      <td>1747.0</td>\n",
       "      <td>0.553531</td>\n",
       "      <td>0.337403</td>\n",
       "      <td>[GCF_000145115.1] Cutibacterium acnes HL030PA1...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 pt_gambit_dist  pt_gambit_sp_thresh  \\\n",
       "jared_id                                               \n",
       "16AC1611140BCAP           0.337                0.554   \n",
       "\n",
       "                                   pt_gambit_genus pt_gambit_species  \\\n",
       "jared_id                                                               \n",
       "16AC1611140BCAP  Propionibacterium (Cutibacterium)             acnes   \n",
       "\n",
       "                                             pt_reported             pt_cap  \\\n",
       "jared_id                                                                      \n",
       "16AC1611140BCAP  Propionibacterium (Cutibacterium) acnes  Propionibacterium   \n",
       "\n",
       "                                     pt_gambit_predicted   new_predicted.name  \\\n",
       "jared_id                                                                        \n",
       "16AC1611140BCAP  Propionibacterium (Cutibacterium) acnes  Cutibacterium acnes   \n",
       "\n",
       "                new_predicted.rank  new_predicted.ncbi_id  \\\n",
       "jared_id                                                    \n",
       "16AC1611140BCAP            species                 1747.0   \n",
       "\n",
       "                 new_predicted.threshold  new_closest.distance  \\\n",
       "jared_id                                                         \n",
       "16AC1611140BCAP                 0.553531              0.337403   \n",
       "\n",
       "                                           new_closest.description  \\\n",
       "jared_id                                                             \n",
       "16AC1611140BCAP  [GCF_000145115.1] Cutibacterium acnes HL030PA1...   \n",
       "\n",
       "                new_next.name new_next.rank  new_next.ncbi_id  \\\n",
       "jared_id                                                        \n",
       "16AC1611140BCAP           NaN           NaN               NaN   \n",
       "\n",
       "                 new_next.threshold  \n",
       "jared_id                             \n",
       "16AC1611140BCAP                 NaN  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "comp[~matches]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5f466b4-5d50-46d5-8127-70a3a62820e9",
   "metadata": {},
   "source": [
    "## Write output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a4ced33d-045b-4821-9093-8f1e49ce6ef8",
   "metadata": {},
   "outputs": [],
   "source": [
    "comp.to_csv(outfiles['comp'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Gambit Publication",
   "language": "python",
   "name": "gambit-pub"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
